hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -7590,41 +7590,45 @@ private void handleTypeComparison(
 		final TableGroup tableGroup = getFromClauseIndex().getTableGroup( typeExpression.getNavigablePath().getParent() );
 		final EntityMappingType entityMappingType = (EntityMappingType) tableGroup.getModelPart().getPartMappingType();
 		if ( entityMappingType.getDiscriminatorMapping().hasPhysicalColumn() ) {
-			// Prevent pruning of the root type's table reference containing the physical discriminator column
+			// If the entity has a physical discriminator column we don't need to register any FILTER usages.
+			// Register only an EXPRESSION usage to prevent pruning of the root type's table reference which
+			// contains the physical discriminator column
 			registerEntityNameUsage(
 					tableGroup,
 					EntityNameUse.EXPRESSION,
 					entityMappingType.getRootEntityDescriptor().getEntityName()
 			);
 		}
-		if ( literalExpressions == null ) {
-			// We have to assume all types are possible and can't do optimizations
-			registerEntityNameUsage( tableGroup, EntityNameUse.FILTER, entityMappingType.getEntityName() );
-			for ( EntityMappingType subMappingType : entityMappingType.getSubMappingTypes() ) {
-				registerEntityNameUsage( tableGroup, EntityNameUse.FILTER, subMappingType.getEntityName() );
-			}
-		}
 		else {
-			if ( inclusive ) {
-				for ( EntityTypeLiteral literalExpr : literalExpressions ) {
-					registerEntityNameUsage(
-							tableGroup,
-							EntityNameUse.FILTER,
-							literalExpr.getEntityTypeDescriptor().getEntityName()
-					);
+			if ( literalExpressions == null ) {
+				// We have to assume all types are possible and can't do optimizations
+				registerEntityNameUsage( tableGroup, EntityNameUse.FILTER, entityMappingType.getEntityName() );
+				for ( EntityMappingType subMappingType : entityMappingType.getSubMappingTypes() ) {
+					registerEntityNameUsage( tableGroup, EntityNameUse.FILTER, subMappingType.getEntityName() );
 				}
 			}
-			else {
-				final Set<String> excludedEntityNames = new HashSet<>( entityMappingType.getSubMappingTypes().size() );
-				for ( EntityTypeLiteral literalExpr : literalExpressions ) {
-					excludedEntityNames.add( literalExpr.getEntityTypeDescriptor().getEntityName() );
-				}
-				if ( !excludedEntityNames.contains( entityMappingType.getEntityName() ) ) {
-					registerEntityNameUsage( tableGroup, EntityNameUse.FILTER, entityMappingType.getEntityName() );
+			if ( literalExpressions != null ) {
```

Comment:
Minor, but this is just the `else` branch to the `if ( literalExpressions == null )` check.  No need for another check.

Reply:
Oops, must have changed that by mistake - thanks!